torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -3496,62 +3500,91 @@ def full_optim_state_dict(
             then nonzero ranks return an empty :class:`dict`.
         """
         osd = optim.state_dict()
-        osd_state, osd_param_groups = osd["state"], osd["param_groups"]  # alias
-
-        group = model.process_group if hasattr(model, "process_group") \
-            else None  # not all `torch.nn.Module`s have `process_group`
+        osd_state, osd_param_groups = osd["state"], osd["param_groups"]
+        process_groups = set(
+            module.process_group for module in model.modules()
+            if hasattr(module, "process_group")
+        )  # see if any modules have a process group attribute
+        if len(process_groups) == 0:
+            group = None  # use default group
+        else:
+            if len(process_groups) > 1:
+                warnings.warn(
```

Comment:
Refer to the comment above about multiple process group support.

Reply:
Can it break in unexpected ways if this is the case? If so, we should probably just be safe and raise an error here?